### PR TITLE
Maglev: Compact maglev table

### PR DESCRIPTION
**Pull request by KBaichoo**
Commit Message: Implement a compact maglev table, and pick between the implementations based on the cost.
Additional Description:
Estimated savings are as follows using default maglev table size:
Screenshot 2023-01-26 10 20 23 AM

Running existing benchmark: 
```
bazel test -c opt //test/common/upstream:load_balancer_benchmark_test
```
With Original Maglev:
```
-----------------------------------------------------------------------------
[2023-01-26 16:21:17.582][19][critical] [test/benchmark/main.cc:103] Expensive benchmarks are being skipped; see test/README.md for more information
2023-01-26T16:21:17+00:00
Running /b/f/w/bazel-out/k8-opt/bin/test/common/upstream/load_balancer_benchmark_test.runfiles/envoy/test/common/upstream/load_balancer_benchmark
Run on (2 X 2800.28 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x1)
  L1 Instruction 32 KiB (x1)
  L2 Unified 1024 KiB (x1)
  L3 Unified 33792 KiB (x1)
Load Average: 0.15, 0.12, 0.04
----------------------------------------------------------------------------------------------------------------------
Benchmark                                                            Time             CPU   Iterations UserCounters...
----------------------------------------------------------------------------------------------------------------------

benchmarkMaglevLoadBalancerBuildTable/100                         4.86 ms         4.87 ms            1 memory=1057.22k memory_per_host=10.572k
benchmarkMaglevLoadBalancerBuildTable/200                         4.97 ms         4.97 ms            1 memory=1057.22k memory_per_host=5.286k
benchmarkMaglevLoadBalancerBuildTable/500                         5.42 ms         5.42 ms            1 memory=1057.22k memory_per_host=2.114k

benchmarkMaglevLoadBalancerChooseHost/100/100000                  9.79 ms         9.79 ms            1 mean_hits=1000 relative_stddev_hits=0.029227 stddev_hits=29.227
benchmarkMaglevLoadBalancerChooseHost/200/100000                  10.6 ms         10.6 ms            1 mean_hits=500 relative_stddev_hits=0.0415841 stddev_hits=20.7921
benchmarkMaglevLoadBalancerChooseHost/500/100000                  12.1 ms         12.1 ms            1 mean_hits=200 relative_stddev_hits=0.0687299 stddev_hits=13.746

benchmarkMaglevLoadBalancerHostLoss/500/1/10000                   16.9 ms         16.8 ms            1 host_loss_over_N_optimal=0.2 percent_different=1
benchmarkMaglevLoadBalancerHostLoss/500/2/10000                   16.8 ms         16.7 ms            1 host_loss_over_N_optimal=0.4 percent_different=1.65
benchmarkMaglevLoadBalancerHostLoss/500/3/10000                   16.4 ms         16.3 ms            1 host_loss_over_N_optimal=0.6 percent_different=1.9
benchmarkMaglevLoadBalancerWeighted/500/5/1/1/10000               16.6 ms         16.6 ms            1 optimal_percent_different=0 percent_different=0

benchmarkMaglevLoadBalancerWeighted/500/5/1/127/1000              15.9 ms         15.8 ms            1 optimal_percent_different=81.9863 percent_different=80.8
benchmarkMaglevLoadBalancerWeighted/500/5/127/1/10000             19.7 ms         19.6 ms            1 optimal_percent_different=81.9863 percent_different=81.72
benchmarkMaglevLoadBalancerWeighted/500/50/1/127/1000             14.8 ms         14.8 ms            1 optimal_percent_different=49.2188 percent_different=50.8
benchmarkMaglevLoadBalancerWeighted/500/50/127/1/10000            16.9 ms         16.9 ms            1 optimal_percent_different=49.2188 percent_different=50.16
benchmarkMaglevLoadBalancerWeighted/500/95/1/127/1000             14.5 ms         14.4 ms            1 optimal_percent_different=4.95857 percent_different=6.5
benchmarkMaglevLoadBalancerWeighted/500/95/127/1/10000            16.5 ms         16.5 ms            1 optimal_percent_different=4.95857 percent_different=7.25
benchmarkMaglevLoadBalancerWeighted/500/95/25/75/1000             15.2 ms         15.2 ms            1 optimal_percent_different=0.139958 percent_different=0.8
benchmarkMaglevLoadBalancerWeighted/500/95/75/25/10000            16.9 ms         16.9 ms            1 optimal_percent_different=0.139958 percent_different=0.74
```

With Compact Maglev:
```
CPU Caches:
  L1 Data 32 KiB (x1)
  L1 Instruction 32 KiB (x1)
  L2 Unified 1024 KiB (x1)
  L3 Unified 33792 KiB (x1)
Load Average: 0.17, 0.10, 0.04
----------------------------------------------------------------------------------------------------------------------
Benchmark                                                            Time             CPU   Iterations UserCounters...
----------------------------------------------------------------------------------------------------------------------
benchmarkMaglevLoadBalancerBuildTable/100                         4.14 ms         4.13 ms            1 memory=67.816k memory_per_host=678
benchmarkMaglevLoadBalancerBuildTable/200                         4.20 ms         4.19 ms            1 memory=77.416k memory_per_host=387
benchmarkMaglevLoadBalancerBuildTable/500                         4.69 ms         4.60 ms            1 memory=90.6k memory_per_host=181

benchmarkMaglevLoadBalancerChooseHost/100/100000                  8.75 ms         8.75 ms            1 mean_hits=1000 relative_stddev_hits=0.029227 stddev_hits=29.227
benchmarkMaglevLoadBalancerChooseHost/200/100000                  9.41 ms         9.41 ms            1 mean_hits=500 relative_stddev_hits=0.0415841 stddev_hits=20.7921
benchmarkMaglevLoadBalancerChooseHost/500/100000                  10.6 ms         10.6 ms            1 mean_hits=200 relative_stddev_hits=0.0687299 stddev_hits=13.746

benchmarkMaglevLoadBalancerHostLoss/500/1/10000                   14.1 ms         14.0 ms            1 host_loss_over_N_optimal=0.2 percent_different=1
benchmarkMaglevLoadBalancerHostLoss/500/2/10000                   13.9 ms         13.8 ms            1 host_loss_over_N_optimal=0.4 percent_different=1.65
benchmarkMaglevLoadBalancerHostLoss/500/3/10000                   13.3 ms         13.3 ms            1 host_loss_over_N_optimal=0.6 percent_different=1.9

benchmarkMaglevLoadBalancerWeighted/500/5/1/1/10000               13.8 ms         13.8 ms            1 optimal_percent_different=0 percent_different=0
benchmarkMaglevLoadBalancerWeighted/500/5/1/127/1000              14.3 ms         14.2 ms            1 optimal_percent_different=81.9863 percent_different=80.8
benchmarkMaglevLoadBalancerWeighted/500/5/127/1/10000             15.7 ms         15.6 ms            1 optimal_percent_different=81.9863 percent_different=81.72
benchmarkMaglevLoadBalancerWeighted/500/50/1/127/1000             12.3 ms         12.2 ms            1 optimal_percent_different=49.2188 percent_different=50.8
benchmarkMaglevLoadBalancerWeighted/500/50/127/1/10000            13.4 ms         13.4 ms            1 optimal_percent_different=49.2188 percent_different=50.16
benchmarkMaglevLoadBalancerWeighted/500/95/1/127/1000             11.6 ms         11.5 ms            1 optimal_percent_different=4.95857 percent_different=6.5
benchmarkMaglevLoadBalancerWeighted/500/95/127/1/10000            13.0 ms         13.0 ms            1 optimal_percent_different=4.95857 percent_different=7.25
benchmarkMaglevLoadBalancerWeighted/500/95/25/75/1000             11.5 ms         11.5 ms            1 optimal_percent_different=0.139958 percent_different=0.8
benchmarkMaglevLoadBalancerWeighted/500/95/75/25/10000            13.2 ms         13.2 ms            1 optimal_percent_different=0.139958 percent_different=0.74

```
In particular I would like to point out the change in memory usage in `benchmarkMaglevLoadBalancerBuildTable`  between the two. I suspect the reason the compact maglev is performing better than the original maglev is due to L1 data cache sizes  -- in practice in a deployment (e.g. load testing QPS the Maglev), I expect both should be similar, but given that Maglev choosing a host is not a cpu bottleneck (and we can easily turn it off).

Risk Level: low-medium, release flag would make it easy to default to original behavior. 
Testing: Added tests
Docs Changes:
Release Notes: Included
Platform Specific Features: `BitArray` does not support windows. 
Runtime guard: Yes

### CODE REVIEW

File: source/common/common/bit_array.h
Comment (pradeepcrao) [inline source/common/common/bit_array.h:114]: Why not use `std::vector`, since it'll only be an additional 8 bytes?

**Review by pradeepcrao — COMMENTED** [review]
Thanks so much for doing this! Very nifty idea :-)

File: source/common/common/bit_array.h
Comment (KBaichoo) [reply source/common/common/bit_array.h:33]: They're orthogonal (will comment the ctor for this).

For ex: we might have 10000 1 bit sized elements (e.g. this is a bitmap) or we might have 10000 7 bit sized elements.

File: source/common/common/bit_array.h
Comment (KBaichoo) [reply source/common/common/bit_array.h]: done.

File: source/common/common/bit_array.h
Comment (KBaichoo) [reply source/common/common/bit_array.h:126]: Makes sure I alloc the exact amount of memory without having to do any shrink to fit or similar dance. 

The vec for 10k might stl vec have size != capacity and alloc additional space.

File: source/common/common/bit_array.h
Comment (KBaichoo) [reply source/common/common/bit_array.h]: done.

File: test/common/common/bit_array_test.cc
Comment (KBaichoo) [reply test/common/common/bit_array_test.cc:23]: It will change depending on the *SAN. Just leaving empty to keep it simple.

**Review by KBaichoo — COMMENTED** [review]
Thanks for the review @pradeepcrao !

File: source/common/common/bit_array.h
Comment (pradeepcrao) [reply source/common/common/bit_array.h:33]: That makes sense. I was being daft.

**Review by pradeepcrao — COMMENTED** [review]
The asan failure looks legit.

What do you think about adding a performance test to see how this compares to std::vector?

File: test/common/common/bit_array_test.cc
Comment (pradeepcrao) [inline test/common/common/bit_array_test.cc:41]: Should this be `bit_width_` ? I don't see where `width` is defined.

File: source/common/upstream/maglev_lb.cc
Comment (pradeepcrao) [inline source/common/upstream/maglev_lb.cc:154]: It seems a little wonky to me that a base class method should know about subclasses. WDYT about making this a standalone method?

File: source/common/upstream/maglev_lb.h
Comment (pradeepcrao) [inline source/common/upstream/maglev_lb.h:145]: There's seems a fair amount of code duplicated between the two derived classes. 

What do you think about having an abstraction around just `table_` and `host_table_` for `CompactMaglevTable` and `table_` for `OriginalMaglevTable`?

File: source/common/common/bit_array.h
Comment (pradeepcrao) [inline source/common/common/bit_array.h:93]: What do you think about padding the allocated space by `WordSize` so we don't need the else code block? Either way is fine with me, but doing so will make the code easier to follow.

Also, could you please add more comments? The code here is very hard for me to follow/reason about.

File: source/common/upstream/maglev_lb.h
Comment (pradeepcrao) [inline source/common/upstream/maglev_lb.h:49]: nit: s/const/constexpr

File: source/common/upstream/maglev_lb.cc
Comment (pradeepcrao) [inline source/common/upstream/maglev_lb.cc:17]: Is this `sizeof(std::shared_ptr<T>) * table_size` ?

File: source/common/upstream/maglev_lb.cc
Comment (pradeepcrao) [inline source/common/upstream/maglev_lb.cc:168]: Since most of the code here is repeated for both original and compact maglev tables, what do you think about my previous comment about wrapping just `table_` and `host_table_` in a class for compact representation, and `table_` from the original maglev table for the "original" representation? 

Will need a solution for tracking "occupied" slots though. You could just keep the `occupied` object in the combined implementation of the algorithm.

Comment (pradeepcrao) [reply source/common/upstream/maglev_lb.cc:168]: @jmarantz What do you think? (since you've been tagged as a reviewer).

File: source/common/upstream/maglev_lb.h
Comment (pradeepcrao) [reply source/common/upstream/maglev_lb.h:138]: I was unclear with my wording. I was suggesting that instead of duplicating the algorithmic details in both derived classes, limit the functionality of the derived classes to just setting/getting the host for an index.

File: source/common/upstream/maglev_lb.cc
Comment (pradeepcrao) [reply source/common/upstream/maglev_lb.cc:168]: Based on my discussion with @KBaichoo offline, I'm okay with having a TODO for now to evaluate this again at a later date.

File: source/common/common/bit_array.h
Comment (KBaichoo) [reply source/common/common/bit_array.h:93]: done.

File: source/common/upstream/maglev_lb.cc
Comment (KBaichoo) [reply source/common/upstream/maglev_lb.cc:168]: sg, added a TODO

**Review by KBaichoo — COMMENTED** [review]
Thanks for the review @pradeepcrao , I think the bit array looks much simpler now with the padding.

File: source/common/upstream/maglev_lb.cc
Comment (pradeepcrao) [inline source/common/upstream/maglev_lb.cc:17]: nit: `constexpr`

**Review by pradeepcrao — COMMENTED** [review]
LGTM.

Thanks so much for working on this @KBaichoo !

**Review by pradeepcrao — DISMISSED** [review]

**Review by jmarantz — COMMENTED** [review]
flushing comments

File: source/common/common/bit_array.h
Comment (jmarantz) [reply source/common/common/bit_array.h:47]: Damn. For some reason I had thought it was. However stackoverflow has found a few options, which I messed around with:
```
#include <cstdio>
#include <cstdint>

#if _WIN32 || _WIN64
    #if _WIN64
        #define POINTER_SIZE 8
    #else
        #define POINTER_SIZE 4
    #endif
#elif __GNUC__
    #if __x86_64__ || __ppc64__
        #define POINTER_SIZE 8
    #else
        #define POINTER_SIZE 4
    #endif
#elif UINTPTR_MAX > UINT_MAX
    #define POINTER_SIZE 8
#else
    #define POINTER_SIZE 4
#endif


#define IMAX_BITS(m) ((m)/((m)%255+1) / 255%255*8 + 7-86/((m)%255+12))
#define UINTPTR_MAX_BITWIDTH IMAX_BITS(UINTPTR_MAX)

int main() {
  fprintf(stderr, "uintpr_max_pointer size=%d\n", POINTER_SIZE);
  fprintf(stderr, "uintpr_max_bitwidth=%ld\n", UINTPTR_MAX_BITWIDTH);
  fprintf(stderr, "uintpr_ptrmax=%lx\n", UINTPTR_MAX);
}
```
yields
```
uintpr_max_pointer size=8
uintpr_max_bitwidth=64
uintpr_ptrmax=ffffffffffffffff
```

hooking on little-endian seems wrong.

https://stackoverflow.com/questions/4079243/how-can-i-use-sizeof-in-a-preprocessor-macro

There were other hacks on the web also. But ultimately it's not too hard to do static-assert but I think it's cleaner to do correct by construction, which you can do by basically depending on compiler-supplied preprocessor macros, which seems OK. I'll let you sort out which mechanism seems most robust and maintainable.

File: source/common/common/bit_array.h
Comment (jmarantz) [inline source/common/common/bit_array.h:95]: nit: you could probably get away (performance wise) with factoring out the common parts of this with get() into common helper function(s). It should all be inlined by the compiler anyway.

**Review by jmarantz — COMMENTED** [review]
this is great stuff -- a few more comments. 

LGTM otherwise.

**Comment by jmarantz** [comment]
Also I think a senior maintainer should review after this that's pretty familiar with the LBs. probably @snowp or @mattklein123

File: source/common/upstream/maglev_lb.h
Comment (jmarantz) [inline source/common/upstream/maglev_lb.h:140]: I have another suggestion which might further improve performance and reduce memory. Maybe for a follow-up.

At a high level, this could be a variable-size data structure. See `InlineStorage` in source/common/common/utility.h. It might break the abstraction between BitArray and CompactMaglevTable so it would be a little messy.

But at the end you'd have one less indirection and 8 fewer bytes: the unique_ptr in BitArray would not be needed as the data would be inline.

Comment (KBaichoo) [reply source/common/upstream/maglev_lb.h:140]: Cool to learn about `InlineStorage` as an available technique. I think here the odds of it helping might be slim as the size of the BitArray object itself would balloon in size and wouldn't fit in cache lines. Which would likely lead to an additional memory access. We can revisit if this becomes an issue.

File: source/common/common/bit_array.h
Comment (KBaichoo) [reply source/common/common/bit_array.h:43]: You're right re: endianess, removed. 

Only requirement is being 8 byte pointers, I think a lot of the codebase I think assumes 64-bit, so having a static_assert which is a C++ standard seems to be a more maintainable way to state this assumption than toolchain specific elements.

File: source/common/common/bit_array.h
Comment (KBaichoo) [reply source/common/common/bit_array.h:91]: Leaning towards keeping the code where it is so it can flow better with the comments, context.

**Review by KBaichoo — COMMENTED** [review]
Thank you for the very through review @jmarantz , learned some new stuff 🚀

File: source/common/common/bit_array.h
Comment (jmarantz) [reply source/common/common/bit_array.h:43]: You're probably right; static_assert seems fine now that you've removed ifdefing altogether. Matt will let us know if e-m needs 32-bit. I'm guessing that it does not.

File: source/common/common/bit_array.h
Comment (jmarantz) [reply source/common/common/bit_array.h:91]: ok, sgtm

File: source/common/common/bit_array.h
Comment (KBaichoo) [reply source/common/common/bit_array.h:40]: The factor I think that helped with the microbenchmark is that the compact table representation e.g. `array_start_` can have much more data fit in L1 data cache (32kb) than the original table representation.

File: source/common/common/bit_array.h
Comment (mattklein123) [inline source/common/common/bit_array.h:27]: This is probably a dumb question but isn't this https://en.cppreference.com/w/cpp/utility/bitset? If it's not can you add comments on how it's different?

**Review by mattklein123 — COMMENTED** [review]
One quick question to start out with. Cool stuff!

File: source/common/common/bit_array.h
Comment (KBaichoo) [reply source/common/common/bit_array.h:27]: From comment above: " BitArray is an fixed sized array of fixed width bits."

Note the "fixed width bits", entries can range from being 1-32 bits each. For bit set elements are just 1 bit width. From looking at get: https://en.cppreference.com/w/cpp/utility/bitset/operator_at we can see it just rets a bool

Comment (mattklein123) [reply source/common/common/bit_array.h:27]: OK, got it. Pedantically I found this a bit confusing. Maybe something like "BitArray is a fixed sized array of fixed with bit sets" or something like that, or just a note to clarify how it's different from bitset. As an aside, I assume that in some cases `std::array<std::bitset>` or similar would not compress as well, but would that end up a simpler implementation? Anyway I would just have some note about bitset to make it more clear for quick readers.

Comment (KBaichoo) [reply source/common/common/bit_array.h:27]: sg, added a note to clarify how it is different from `std::bitset`.

Yea the compression is __significantly worse__ in all cases except when using bit array for 32-bit elements.
```
std::bitset<1> foo;
std::cout << "Size of foo: " << sizeof(foo) << std::endl;
```

Yields 4, which would be 32x  more than BitArray representation. Moreover we'd need to instantiate templates for `bitset<1>, bitset<2>, ..., bitset<32>`. which seems like an easy way to bloat the binary.

**Review by mattklein123 — APPROVED** [review]
Nice. LGTM at a high level!

**Comment by jpsim** [comment]
This looks neat @KBaichoo, although I think this caused a build failure for Envoy Mobile for some architectures:

```
source/common/common/bit_array.h:48:5: error: static_assert failed due to requirement 'sizeof(unsigned char *) == 8' "Pointer underlying size is not 8 bytes, are we running on a 64-bit architecture?"
    static_assert(
    ^
```

We build for the following architectures: `x86`, `x86_64`, `armeabi-v7a`, `arm64-v8a`

I suspect it's the `x86` that's failing based on the error message.

**Comment by jmarantz** [comment]
Ah we had asserted that E-M did not every run on 32-bit. If that's not the case I think we can roll this back and then push forward with this PR, but with https://github.com/envoyproxy/envoy/pull/25149/commits/1df41955766a197df2eb41544d04f0131dfede2e backed out, and replaced with code that ifdefs based on bit-width

I added comments suggesting how we might do that, as it's not possible to us sizeof in an ifdef.

@KBaichoo WDYT?

**Comment by jmarantz** [comment]
Comment with suggestions about how to ifdef for bit-size: https://github.com/envoyproxy/envoy/pull/25149/files/1df41955766a197df2eb41544d04f0131dfede2e#r1103639479

**Comment by KBaichoo** [comment]
That sounds a reasonable approach @jmarantz . Unless there's a better way that EM approaches these e.g. I suspect there might be other parts in the codebase that originally assumed 64-bit?

@jpsim can you provide a bazel command so I could try to test this in the next iteration?

**Comment by jpsim** [comment]
This should trigger it:

```
cd mobile
./bazelw build --fat_apk_cpu=x86_64 //:android_dist
```

**Comment by jmarantz** [comment]
I would've thought x86_64 would have passed that static assert. Did you mean --fat_apk_cpu=x86 ?

**Comment by jpsim** [comment]
Yes, sorry I typo'd. I meant `--fat_apk_cpu=x86`.

**Pull request merged**
